server/src/test/scala/com/cloudera/livy/utils/SparkYarnAppSpec.scala

```diff
@@ -0,0 +1,197 @@
+/*
+ * Licensed to Cloudera, Inc. under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  Cloudera, Inc. licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.cloudera.livy.utils
+
+import java.util.concurrent.{CountDownLatch, TimeUnit}
+
+import scala.concurrent.duration._
+import scala.language.postfixOps
+import scala.util.control.Exception.noCatch
+
+import org.apache.hadoop.yarn.api.records.{ApplicationId, ApplicationReport, FinalApplicationStatus, YarnApplicationState}
+import org.apache.hadoop.yarn.api.records.FinalApplicationStatus.UNDEFINED
+import org.apache.hadoop.yarn.api.records.YarnApplicationState._
+import org.apache.hadoop.yarn.client.api.YarnClient
+import org.apache.hadoop.yarn.util.ConverterUtils
+import org.mockito.Mockito._
+import org.mockito.invocation.InvocationOnMock
+import org.mockito.stubbing.Answer
+import org.scalatest.FunSpec
+import org.scalatest.mock.MockitoSugar.mock
+
+import com.cloudera.livy.LivyConf
+import com.cloudera.livy.util.LineBufferedProcess
+import com.cloudera.livy.utils.SparkApp._
+
+class SparkYarnAppSpec extends FunSpec {
+  private def cleanupThread(t: Thread)(f: => Unit) = {
+    noCatch.andFinally { t.interrupt() } (f)
+  }
+
+  private def mockSleep(ms: Long) = {
+    Thread.`yield`()
+  }
+
+  private def withSleepMethod(mockSleep: Long => Unit)(f: => Unit): Unit = {
+    noCatch.andFinally { Clock.setSleepMethod(Thread.sleep) } {
```

Comment:
I don't see why it defeats anything. You're adding this clock implementation because *specific* unit tests need it. But, if you do what you did before and forget to clean up, you'll affect *all* unit tests that run after yours.

The interface I'm proposing has exactly the same functionality you seek, but it avoids the problem of forgetting to reset the clock for other tests. So I don't see why it defeats anything.